### PR TITLE
feat(devtools): Implement Devtools usage telemetry

### PR DESCRIPTION
**Pull request by WayneFerrao**
AB#7119

##Description

This draft serves as a working proof of concept for sending Devtools usage telemetry data to Kusto through the introduction of the sessionID and continuityID identifiers.

### CODE REVIEW

File: packages/tools/devtools/devtools-view/src/TelemetryUtils.ts [hunk 1]
```diff
@@ -42,9 +44,17 @@ export function useLogger(): ITelemetryLoggerExt | undefined {
...
+	private readonly continuityID?: string;
+	public constructor(private readonly baseLogger?: ITelemetryBaseLogger) {
+		this.sessionID = uuidv4();
+		this.continuityID = getOrCreateBrowserId();
+	}
 
 	public send(event: ITelemetryBaseEvent): void {
+		event.sessionID = this.sessionID;
+		event.continuityID = this.continuityID;
+
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-view/src/TelemetryUtils.ts:57]: I don't think we need this in the console logger, where the users already have some "continuity" to see that all messages written to console are related to each other.

File: packages/utils/telemetry-utils/src/telemetryTypes.ts [hunk 2]
```diff
@@ -78,6 +78,8 @@ export interface ITelemetryEventExt extends ITelemetryPropertiesExt {
 export interface ITelemetryGenericEventExt extends ITelemetryPropertiesExt {
 	eventName: string;
 	category?: TelemetryEventCategory;
+	sessionID?: string | undefined;
+	continuityID?: string | undefined;
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/src/telemetryTypes.ts:82]: We cannot put these here, this is a very general interface for telemetry and we should not force every telemetry message that Fluid generates to have these properties :). But it shouldn't be necessary, we can already add arbitrary key/value pairs to the events.

**Review by alexvy86 — COMMENTED** [review]
The idea is on the right track, but we need to do it in a different place. devtools-view is _technically_ reusable, even if we don't intend it that way for now. But the devtools-example app reuses it to display the FluidDevtools UI inline with the app, instead of needing the actual browser extension installed. In that scenario, we don't want telemetry to be submitted. Only the extension should. So the correct place to make changes is close to [here](https://github.com/microsoft/FluidFramework/blob/4c06310956f44fe30e21012a42b1f52c885f23f8/packages/tools/devtools/devtools-browser-extension/src/devtools/InitializeView.ts#L32-L38), where the extension passes a logger (a class we implemented) to a component from devtools-view. The send() method of that logger already adds some fields to the events that go through it. We need to add more `Data.<propertyName>` entries for the new fields.

File: packages/tools/devtools/devtools-view/src/TelemetryUtils.ts (hunk 1 above)
Comment (WayneFerrao) [reply packages/tools/devtools/devtools-view/src/TelemetryUtils.ts:57]: Should these identifiers be removed entirely from the console logger?

Comment (alexvy86) [reply packages/tools/devtools/devtools-view/src/TelemetryUtils.ts]: Yeah, in general I don't think we need any changes in the devtools-view package to implement this.

File: packages/tools/devtools/devtools-view/src/TelemetryContext.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+export const getOrCreateContinuityID = (): string => {
+	let continuityID = localStorage.getItem(CONTINUITY_ID_KEY);
+
+	if (continuityID === null || continuityID === "") {
+		continuityID = uuidv4();
+		localStorage.setItem(CONTINUITY_ID_KEY, continuityID);
+	}
+
+	return continuityID;
+};
```
Comment (seanimam) [inline packages/tools/devtools/devtools-view/src/TelemetryContext.ts:23]: Is there a reason this function is within its own file? If the logic will only be used in the constructor for `ConsoleVerboseLogger` then I would consider just keeping the logic within the constructor.

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 4]
```diff
@@ -70,12 +71,19 @@ export class OneDSLogger implements ITelemetryBaseLogger {
 	 */
 	private readonly enabled: boolean = false;
 
+	private readonly sessionID?: string;
+	private readonly continuityID?: string;
+	private readonly CONTINUITY_ID_KEY = "myAppBrowserId";
```
Comment (alexvy86) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:76]: A bit to follow the naming convention we use for other flags/keys, and to complete the move away from the potentially misleading "browserId".
```suggestion
	private readonly CONTINUITY_ID_KEY = "Fluid.Devtools.ContinuityId";
```

File: packages/utils/telemetry-utils/api-report/telemetry-utils.api.md [hunk 5]
```diff
@@ -219,7 +219,11 @@ export interface ITelemetryGenericEventExt extends ITelemetryPropertiesExt {
     // (undocumented)
     category?: TelemetryEventCategory;
     // (undocumented)
+    continuityID?: string | undefined;
+    // (undocumented)
     eventName: string;
+    // (undocumented)
+    sessionID?: string | undefined;
```
Comment (alexvy86) [inline packages/utils/telemetry-utils/api-report/telemetry-utils.api.md:226]: I imagine the PR build will report an unexpected change in this file. Need to rebuild this package to undo the changes to its public API surface.

**Review by alexvy86 — APPROVED** [review]
One of the comments below should cause an error in the build so we'll have to fix it. Small change in the other, and 🚀 after that :).

**Review by seanimam — APPROVED** [review]
Looks good to me

**Pull request merged**